### PR TITLE
Properly warn if food will cause you to overeat

### PR DESCRIPTION
**Pull request by RenechCDDA**
#### Summary
Bugfixes "Properly warn if food will cause you to overeat"

#### Purpose of change
Same reason as #74149, make camp meals not stuff your face without warning

Character::will_eat and Character::update_stomach had different ideas about when the stomach was full

#### Describe the solution
Take them out to a nice foodplace™ dinner so they can sort out their differences (new function, both of them use it. one source of truth)

Also updated the camp menu while I was here so you can choose how you stuff your piehole. Please eat responsibly, even ingame.

#### Describe alternatives you've considered
`units::volume in_stomach_volume = food.volume() * compute_effective_food_volume_ratio( food );`
I bet there's other places that should be using this, I just haven't found them yet

#### Testing
(After first commit, before second commit):
https://github.com/CleverRaven/Cataclysm-DDA/assets/84619419/92ada55c-5231-47ab-bbf1-4cc265617d18

### CODE REVIEW

**Comment by RenechCDDA** [comment]
...I have no idea how that test failure has happened, but it seems like it may be related. Let's roll the dice (re-run the test) anyway!

**Comment by Kamejeir** [comment]
image
this result from consumption of water via a water tank (with 230 units of water) has me believing that in_stomach_volume takes into account the combined volume of all water available to the faucet, not just the amount expected to be consumed.

by placing only one unit of water (which would fit into a stomach) in the tank, the prompt for being too full disappears. I'm going to try making a local edit to the test with an immediately drinkable amount of water and see if it succeeds

File: src/consumption.cpp [hunk 1]
```diff
@@ -992,8 +992,9 @@ ret_val<edible_rating> Character::will_eat( const item &food, bool interactive )
         add_consequence( _( "Your stomach won't be happy (not rotten enough)." ), ALLERGY_WEAK );
     }
 
+    units::volume in_stomach_volume = food.volume() * compute_effective_food_volume_ratio( food );
```
Comment (Kamejeir) [inline src/consumption.cpp:995]:
```suggestion
    units::volume in_stomach_volume = food.volume( false, false, 1 ) * compute_effective_food_volume_ratio( food );
```
This will solve the test failure. It makes the function not look at the faucet tank as a whole, but only see how much one charge would add. Previously, the test would not drink any water because seeing that it was 8 units (2L, 0.25L per) as a whole had the faucet thinking you were too full to safely consume water, choosing to become greyed-out (not `_enabled`) instead, causing `!stomach_should_be_full` to experience being empty while the faucet was disabled.

Comment (RenechCDDA) [reply src/consumption.cpp:995]: I was a little worried about this because Character::consume variably returns a "All of the item was eaten" or "Some of the item was eaten" value, but it seems that distinction is only for if a charged item lost charges without hitting 0. Basically, we only ever eat one charge at a time, so calculating eaten volume based on 1 charge is objectively correct.

In simpler terms: looks good! 👍
